flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/FlinkFileIO.java

```diff
@@ -0,0 +1,182 @@
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.util.SerializableMap;
+import org.apache.iceberg.util.Tasks;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FlinkFileIO implements FileIO, SupportsPrefixOperations, SupportsBulkOperations {
+  private static final Logger LOG = LoggerFactory.getLogger(FlinkFileIO.class);
+  private static final String DELETE_FILE_PARALLELISM = "iceberg.hadoop.delete-file-parallelism";
+  private static final String DELETE_FILE_POOL_NAME = "iceberg-hadoopfileio-delete";
+  private static final int DELETE_RETRY_ATTEMPTS = 3;
+  private static final int DEFAULT_DELETE_CORE_MULTIPLE = 4;
+  private static volatile ExecutorService executorService;
+  private SerializableMap<String, String> properties = SerializableMap.copyOf(ImmutableMap.of());
+
+  @Override
+  public InputFile newInputFile(String path) {
+    return new FlinkInputFile(new Path(path));
+  }
+
+  @Override
+  public InputFile newInputFile(String path, long length) {
+    return new FlinkInputFile(new Path(path), length);
+  }
+
+  @Override
+  public OutputFile newOutputFile(String path) {
+    return new FlinkOutputFile(new Path(path));
+  }
+
+  @Override
+  public void deleteFile(String path) {
+    Path toDelete = new Path(path);
+    try {
+      toDelete.getFileSystem().delete(toDelete, false /* not recursive */);
+    } catch (IOException e) {
+      throw new UncheckedIOException(String.format("Failed to delete file: %s", path), e);
+    }
+  }
+
+  @Override
+  public Iterable<FileInfo> listPrefix(String prefix) {
+    LOG.debug("Listing {}", prefix);
+    Path prefixToList = new Path(prefix);
+    try {
+      return listPrefix(prefixToList.getFileSystem(), prefixToList);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void deletePrefix(String prefix) {
+    Path prefixToDelete = new Path(prefix);
+
+    try {
+      prefixToDelete.getFileSystem().delete(prefixToDelete, true /* recursive */);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void deleteFiles(Iterable<String> pathsToDelete) throws BulkDeletionFailureException {
+    AtomicInteger failureCount = new AtomicInteger(0);
+    Tasks.foreach(pathsToDelete)
+        .executeWith(executorService())
+        .retry(DELETE_RETRY_ATTEMPTS)
+        .stopRetryOn(FileNotFoundException.class)
+        .suppressFailureWhenFinished()
+        .onFailure(
+            (f, e) -> {
+              LOG.error("Failure during bulk delete on file: {} ", f, e);
+              failureCount.incrementAndGet();
+            })
+        .run(this::deleteFile);
+
+    if (failureCount.get() != 0) {
+      throw new BulkDeletionFailureException(failureCount.get());
+    }
```

Comment:
maybe a `log.debug` statement indicating how many files were deleted from the given prefix

Reply:
I wouldn't put an atomic integer, and synchronization there for just debug logging. Also this is called from multiple places, and without context this info is not that useful